lib/services/watcher.go

```diff
@@ -1468,25 +1519,43 @@ func (n *nodeCollector) resourceKind() string {
 // getResourcesAndUpdateCurrent is called when the resources should be
 // (re-)fetched directly.
 func (n *nodeCollector) getResourcesAndUpdateCurrent(ctx context.Context) error {
-	nodes, err := n.NodesGetter.GetNodes(ctx, apidefaults.Namespace)
+	newCurrent, err := n.getNodes(ctx)
 	if err != nil {
 		return trace.Wrap(err)
 	}
-	defer n.defineCollectorAsInitialized()
+	defer func() {
+		n.defineCollectorAsInitialized()
+		n.stale.Store(false)
```

Comment:
Staleness should only be modified under lock, in order to prevent a concurrent "refresh" from overwriting the node set between the time that the lock is released and this deferred function actually runs.